File: workspace_js/services/js-workflow-executor/src/modules/app/module.ts
import { container } from 'tsyringe';
import type { Logger, Application } from '@mapexos/microservices';
import { LOGGER_TOKEN, APP_TOKEN } from '@mapexos/microservices';

import { registerRoutes } from '@modules/app/interfaces/http';
import { Modules } from '@shared/configuration';

/**
 * InitModule orchestrates all module initializations in phases
 * Following workspace_go pattern:
 * Phase 1: Repositories
 * Phase 2: Services
 * Phase 3: Interfaces (HTTP routes)
 * Phase 4: Listeners (NATS consumers)
 */
export async function initModule(): Promise<void> {
	const logger = container.resolve<Logger>(LOGGER_TOKEN);
	const app = container.resolve<Application>(APP_TOKEN);

	logger.info('[MODULE:APP] Initializing');

	// Register APP module routes
	app.use('/', registerRoutes());

	logger.info('[MODULE:APP] Initialized');
	logger.info({ total: Modules.length, modules: Modules.map(m => m.name) }, '[MODULE:DOMAIN] Initializing domain modules');

	// PHASE 1: Initialize all repositories
	logger.info('[MODULE:PHASE1] Initializing repositories');
	for (const mod of Modules) {
		if (!mod.lazy && mod.initRepositories) {
			logger.debug({ module: mod.name }, '[MODULE:PHASE1] Initializing repositories');
			mod.initRepositories();
		}
	}

	// PHASE 2: Initialize all services
	logger.info('[MODULE:PHASE2] Initializing services');
	for (const mod of Modules) {
		if (!mod.lazy && mod.initServices) {
			logger.debug({ module: mod.name }, '[MODULE:PHASE2] Initializing services');
			mod.initServices();
		}
	}

	// PHASE 3: Initialize all interfaces (HTTP routes)
	logger.info('[MODULE:PHASE3] Initializing interfaces');
	for (const mod of Modules) {
		if (!mod.lazy && mod.initInterfaces) {
			logger.debug({ module: mod.name }, '[MODULE:PHASE3] Initializing interfaces');
			mod.initInterfaces();
		}
	}

	// PHASE 4: Initialize listeners (NATS consumers)
	// Some listeners may be async (e.g., ensureFanoutStream)
	logger.info('[MODULE:PHASE4] Initializing listeners');
	for (const mod of Modules) {
		if (!mod.lazy && mod.initListeners) {
			logger.debug({ module: mod.name }, '[MODULE:PHASE4] Initializing listeners');
			await mod.initListeners();
		}
	}

	logger.info('[MODULE:DOMAIN] All modules initialized');
}

File: workspace_go/services/mapexVault/src/modules/app/module.go
package appModule

import (
	"fmt"

	"github.com/gofiber/fiber/v2"

	logger "github.com/Mapex-Solutions/mapexGoKit/microservices/logger"

	configMod "mapexVault/src/shared/configuration/modules"
)

// InitModule initializes all modules in 3 phases:
// 1. InitRepositories - Registers all repositories in DIG container
// 2. InitServices - Registers all services in DIG container
// 3. InitInterfaces - Registers HTTP routes and consumers
func InitModule(c *fiber.App) {
	logger.Info("[MODULE:App] Initializing Modules")

	for i, mod := range configMod.Modules {
		logger.Info(fmt.Sprintf("[MODULE:App] %d. %s", i+1, mod.Name))
	}

	logger.Info("[MODULE:App] Initializing Repositories")
	for _, mod := range configMod.Modules {
		if !mod.Lazy && mod.InitRepositories != nil {
			logger.Info(fmt.Sprintf("[MODULE:%s] Initializing repositories...", mod.Name))
			mod.InitRepositories()
		}
	}

	logger.Info("[MODULE:App] Initializing Services")
	for _, mod := range configMod.Modules {
		if !mod.Lazy && mod.InitServices != nil {
			logger.Info(fmt.Sprintf("[MODULE:%s] Initializing services...", mod.Name))
			mod.InitServices()
		}
	}

	logger.Info("[MODULE:App] Initializing Interfaces (HTTP Routes)")
	for _, mod := range configMod.Modules {
		if !mod.Lazy && mod.InitInterfaces != nil {
			logger.Info(fmt.Sprintf("[MODULE:%s] Initializing interfaces...", mod.Name))
			mod.InitInterfaces()
		}
	}

	logger.Info("[MODULE:App] All modules initialized successfully")
}
